src/main/scala/io/iohk/ethereum/merklepatriciatrie/MerklePatriciaTrie.scala

```diff
@@ -0,0 +1,486 @@
+        }
+      }
+      else NodeRemoveResult(hasChanged = false, maybeNewChild = Some(node))
+  }
+
+  /**
+    * Given a node which may be in an invalid state, fix it such that it is then in a valid state. Invalid state means:
+    *   - Branch node where there is only a single entry;
+    *   - Extension node followed by anything other than a Branch node.
+    *
+    * @param node         that may be in an invalid state.
+    * @param dataSource   to obtain the nodes referenced in the node that may be in an invalid state.
+    * @param notStoredYet to obtain the nodes referenced in the node that may be in an invalid state,
+    *                     if they were not yet inserted into the dataSource.
+    * @param hashFn
+    * @return fixed node.
+    * @throws MPTException if there is any inconsistency in how the trie is build.
+    */
+  @tailrec
+  private def fix(node: Node, dataSource: DataSource, notStoredYet: Seq[Node], hashFn: HashFn): Node = node match {
+    case BranchNode(children, optStoredValue, _) =>
+      val usedIndexes = children.indices.foldLeft[Seq[Int]](Seq.empty) {
+        (acc, i) =>
+          if (children(i).isDefined) i +: acc else acc
+      }
+      (usedIndexes, optStoredValue) match {
+        case (Nil, None) => throw MPTException("Branch with no subvalues")
+        case (index :: Nil, None) =>
+          val temporalExtNode = ExtensionNode(Array[Byte](index.toByte), children(index).get, hashFn)
+          fix(temporalExtNode, dataSource, notStoredYet, hashFn)
+        case (Nil, Some(value)) => LeafNode(Array.emptyByteArray, value, hashFn)
+        case _ => node
+      }
+    case extensionNode@ExtensionNode(sharedKey, next, _) =>
+      val nextNode = extensionNode.next match {
+        case Left(nextHash) =>
+          // If the node is not in the extension node then it might be a node to be inserted at the end of this remove
+          // so we search in this list too
+          notStoredYet.find(n => n.hash sameElements nextHash).getOrElse(
+            getNextNode(extensionNode, dataSource) // We search for the node in the db
+          )
+        case Right(nextNodeOnExt) => nextNodeOnExt
+      }
+      val newNode = nextNode match {
+        // Compact Two extensions into one
+        case ExtensionNode(subSharedKey, subNext, _) => ExtensionNode(sharedKey ++ subSharedKey, subNext, hashFn)
+        // Compact the extension and the leaf into the same leaf node
+        case LeafNode(subRemainingKey, subValue, _) => LeafNode(sharedKey ++ subRemainingKey, subValue, hashFn)
+        // It's ok
+        case BranchNode(subChildren, subOptValue, _) => node
+      }
+      newNode
+    case _ => node
+  }
+}
+
+private case class NodeInsertResult(newNode: Node,
+                                    toDeleteFromStorage: Seq[Node] = Seq(),
+                                    toUpdateInStorage: Seq[Node] = Seq())
+
+private case class NodeRemoveResult(hasChanged: Boolean, maybeNewChild: Option[Node],
```

Comment:
`NodeInsertResult` and `NodeRemoveResult` are use internally in `MerklePatriciaTree`, right? So maybe it would better to place them in the companion object?